crates/wasm-smith/src/lib.rs

```diff
@@ -584,29 +577,34 @@ where
 
         let num_stable_choices = choices.len();
 
-        arbitrary_loop(u, self.config.max_imports(), |u| {
-            choices.truncate(num_stable_choices);
-            if self.memory_imports() < self.config.max_memories() {
-                choices.push(|u, _| Ok(Import::Memory(u.arbitrary()?)));
-            }
-            if self.table_imports() < self.config.max_tables() {
-                choices.push(|u, m| Ok(Import::Table(m.arbitrary_table_type(u)?)));
-            }
+        arbitrary_loop(
+            u,
+            self.config.min_imports(),
+            self.config.max_imports(),
+            |u| {
+                choices.truncate(num_stable_choices);
```

Comment:
What if `config.min_imports() >= 1` but all of `config.max_{functions,memories,globals,tables}() == 0`? That is, with the introduction of minimums, we can now have contradicting constraints.

I think that is fine so long as we assert that `min <= max` in `arbitrary_loop` and document these requirements for the min/max methods in the `Config` trait (and also document the special case of imports where `min_imports <= sum(max_{functions,memories,globals,tables})`).

Reply:
So I think the simplest resolution here is (as noted above) just count imports as a separate thing with a separate limit; i.e. an imported function does not count against `max_funcs`. If there is some requirement by the user that the total function index space is limited to some range, then that's a different thing, but it seems to me like "locally defined function bodies" is a thing whose range we want to control separately in most cases. I'm not too tied to this, though.

Reply:
This would make things easier however it doesn’t allow us to do things like set max memories to one to avoid the multi-memory proposal

Reply:
Hmm, that's a good point. It appears that the existing implementation wasn't consistent with this (`arbitrary_imports()` is invoked to create imports before `arbitrary_{funcs,tables,memories,globals}()`, and the latter do not check how many imports of each kind have been created); so I'll fix that as well, if these are the semantics that we want.

Reply:
(Actually, it seems only the function case was not checking the number of imports; I was confused by that case. Will fix nevertheless.)